### PR TITLE
instancing support for the backends

### PR DESCRIPTION
**Pull request by pixelflinger**
We just add an instance count to the backend draw command.

### CODE REVIEW

**Review by prideout — APPROVED** [review]

**Review by bejado — APPROVED** [review]

File: android/filament-android/src/main/java/com/google/android/filament/RenderableManager.java [hunk 1]
```diff
@@ -284,6 +284,22 @@ public Builder lightChannel(@IntRange(from = 0, to = 7) int channel, boolean ena
             return this;
         }
 
+        /**
+         * Specifies the number of draw instance if this renderable. The default is 1 instance and
```
Comment (prideout) [inline android/filament-android/src/main/java/com/google/android/filament/RenderableManager.java:288]:
```suggestion
         * Specifies the number of draw instances of this renderable. The default is 1 instance and
```

File: android/filament-android/src/main/java/com/google/android/filament/RenderableManager.java [hunk 2]
```diff
@@ -284,6 +284,22 @@ public Builder lightChannel(@IntRange(from = 0, to = 7) int channel, boolean ena
...
+         * the maximum number of instances allowed is 65535. 0 is invalid.
+         * All instances are culled using the same bounding box, so care must be taken to make
+         * sure all instances render inside the specified bounding box.
+         * The material can use getInstanceId() in the vertex shader to get the instance index and
+         * possibly adjust the position or transform.
+         *
+         * @param instanceCount the number of instance silently clamped between 1 and 65535.
```
Comment (prideout) [inline android/filament-android/src/main/java/com/google/android/filament/RenderableManager.java:295]:
```suggestion
         * @param instanceCount the number of instances silently clamped between 1 and 65535.
```

File: filament/include/filament/RenderableManager.h [hunk 3]
```diff
@@ -328,6 +328,18 @@ class UTILS_PUBLIC RenderableManager : public FilamentAPI {
          */
         Builder& blendOrder(size_t primitiveIndex, uint16_t order) noexcept;
 
+        /**
+         * Specifies the number of draw instance if this renderable. The default is 1 instance and
```
Comment (prideout) [inline filament/include/filament/RenderableManager.h:332]:
```suggestion
         * Specifies the number of draw instances of this renderable. The default is 1 instance and
```

File: filament/include/filament/RenderableManager.h [hunk 4]
```diff
@@ -328,6 +328,18 @@ class UTILS_PUBLIC RenderableManager : public FilamentAPI {
...
+         * the maximum number of instances allowed is 65535. 0 is invalid.
+         * All instances are culled using the same bounding box, so care must be taken to make
+         * sure all instances render inside the specified bounding box.
+         * The material can use getInstanceId() in the vertex shader to get the instance index and
+         * possibly adjust the position or transform.
+         *
+         * @param instanceCount the number of instance silently clamped between 1 and 65535.
```
Comment (prideout) [inline filament/include/filament/RenderableManager.h:339]:
```suggestion
         * @param instanceCount the number of instances silently clamped between 1 and 65535.
```

File: shaders/src/getters.vs [hunk 5]
```diff
@@ -29,6 +29,15 @@ int getVertexIndex() {
 #endif
 }
 
+/** @public-api */
+int getInstanceId() {
```
Comment (bejado) [inline shaders/src/getters.vs:33]: Maybe rename to `getInstanceIndex` to be consistent with `getVertexIndex` above?

**Pull request merged**